contracts/LiquidityConversionRates.sol

```diff
@@ -0,0 +1,265 @@
+            collectedFeesInTwei += calcCollectedFee(abs(buyAmountInTwei));
+        } else {
+            // Sell case
+            collectedFeesInTwei += abs(buyAmountInTwei) * feeInBps / 10000;
+        }
+    }
+
+    event CollectedFeesReset(uint resetFeesInTwei);
+
+    function resetCollectedFees() public onlyAdmin {
+        uint resetFeesInTwei = collectedFeesInTwei;
+        collectedFeesInTwei = 0;
+
+        CollectedFeesReset(resetFeesInTwei);
+    }
+
+    function getRate(
+            ERC20 conversionToken,
+            uint currentBlockNumber,
+            bool buy,
+            uint qtyInSrcWei
+    ) public view returns(uint) {
+
+        currentBlockNumber;
+
+        require(qtyInSrcWei <= MAX_QTY);
+        uint eInFp = fromWeiToFp(reserveContract.balance);
+        uint rateInPrecision = getRateWithE(conversionToken, buy, qtyInSrcWei, eInFp);
+        require(rateInPrecision <= MAX_RATE);
+        return rateInPrecision;
+    }
+
+    function getRateWithE(ERC20 conversionToken, bool buy, uint qtyInSrcWei, uint eInFp) public view returns(uint) {
+        uint deltaEInFp;
+        uint sellInputTokenQtyInFp;
+        uint deltaTInFp;
+        uint rateInPrecision;
+
+        require(qtyInSrcWei <= MAX_QTY);
+        require(eInFp <= maxQtyInFp);
+        if (conversionToken != token) return 0;
+
+        if (buy) {
+            // ETH goes in, token goes out
+            deltaEInFp = fromWeiToFp(qtyInSrcWei);
+            if (deltaEInFp > maxEthCapBuyInFp) return 0;
+
+            if (deltaEInFp == 0) {
+                rateInPrecision = buyRateZeroQuantity(eInFp);
+            } else {
+                rateInPrecision = buyRate(eInFp, deltaEInFp);
+            }
+        } else {
+            sellInputTokenQtyInFp = fromTweiToFp(qtyInSrcWei);
+            deltaTInFp = valueAfterReducingFee(sellInputTokenQtyInFp);
+            if (deltaTInFp == 0) {
+                rateInPrecision = sellRateZeroQuantity(eInFp);
+                deltaEInFp = 0;
+            } else {
+                (rateInPrecision, deltaEInFp) = sellRate(eInFp, sellInputTokenQtyInFp, deltaTInFp);
+            }
+
+            if (deltaEInFp > maxEthCapSellInFp) return 0;
+        }
+
+        rateInPrecision = rateAfterValidation(rateInPrecision, buy);
+        return rateInPrecision;
+    }
+
+    function rateAfterValidation(uint rateInPrecision, bool buy) public view returns(uint) {
+        uint minAllowRateInPrecision;
+        uint maxAllowedRateInPrecision;
+
+        if (buy) {
+            minAllowRateInPrecision = minBuyRateInPrecision;
+            maxAllowedRateInPrecision = maxBuyRateInPrecision;
+        } else {
+            minAllowRateInPrecision = minSellRateInPrecision;
+            maxAllowedRateInPrecision = maxSellRateInPrecision;
+        }
+
+        if ((rateInPrecision > maxAllowedRateInPrecision) || (rateInPrecision < minAllowRateInPrecision)) {
+            return 0;
+        } else if (rateInPrecision > MAX_RATE) {
```

Comment:
yeahh, I separated it since logically we are checking something else.
The 2 checks with the || check the user max rate configs, while the other check is checking for overflow.
leaving as is.